### PR TITLE
Rename parsers to tokenizers, which is more appropriate

### LINKED ISSUE #16: Implement subshell grammar
https://github.com/mgattozzi/ysh/blob/9d6a58d1d5aebd561789c094766ef6605fb4c8c7/src/parse.rs#L175-L183

### PR DESCRIPTION
**Pull request by myrrlyn**
Implement shell, string, and key=val tokenization

Closes #16 in that `$...` sequences are tokenized; actually parsing them is postponed

### CODE REVIEW

File: src/ast/mod.rs [hunk 1]
```diff
@@ -17,7 +21,7 @@ impl<'a> std::iter::Iterator for ArgsIter<'a> {
     type Item = &'a str;
     fn next(&mut self) -> Option<Self::Item> {
         //  Use the tokenizer to get a snippet
-        let (rest, span) = parse::apply(str::trim_left, parse::token)(self.text)
+        let (rest, span) = trim_left(atom)(self.text)
```
Comment (hawkw) [inline src/ast/mod.rs]: TIOLI: I think this reads slightly clearer as `token::atom` still, but 🤷‍♀️

File: src/token.rs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    modifier: impl Fn(T) -> U,
+    tokenizer: impl Fn(U) -> TokenResult<'a, V>,
+) -> impl Fn(T) -> TokenResult<'a, V> {
+    move |input: T| tokenizer(modifier(input))
+}
+
+/// Modifies a tokenizer to call `trim_left` on the input before processing it.
+pub fn trim_left<'a, T>(
+    tokenizer: impl Fn(&'a str) -> TokenResult<'a, T>,
+) -> impl Fn(&'a str) -> TokenResult<'a, T> {
```
Comment (hawkw) [inline src/token.rs]: cool how this now composes without needing to use `compose` :)

File: src/token.rs [hunk 3]
```diff
@@ -0,0 +1,365 @@
+/*! Tokenizers
```
Comment (hawkw) [inline src/token.rs]: Extremely a nit: AFAIK, using block comments is [considered bad style in Rust](https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/guide/guide.md#doc-comments)?

Comment (hawkw) [reply src/token.rs]: That said, I love the amount of comments in this module! Good stuff!

File: src/token.rs (hunk 2 above)
Comment (myrrlyn) [reply src/token.rs]: I realized I was literally only using `compose` in `compose(str::trim_left, whatever)` and ... yup, that deserves a dedicated function.

I am sure I will find other use cases for raw `compose` in the future though.

File: src/token.rs (hunk 3 above)
Comment (myrrlyn) [reply src/token.rs]: That's because the authors of the style guide are *cowards*, but I will change it.

Comment (hawkw) [reply src/token.rs]: If it's _that_ important to you, I don't really care. :)

**Comment by hawkw** [comment]
Sidenote, the AppVeyor CI jobs still seem to be breaking due to #26 --- that is (still) not your fault.

**Comment by myrrlyn** [comment]
Closing in favor of PR #29, which addresses Eliza's "split move and implementation commits" concern

**Pull request closed without merging**